### ISSUE TITLE
Jonathan Hui | RL -- Meta-Learning.

### ISSUE THREAD

**Issue by NorbertZheng** [NorbertZheng/read-papers#66]
Jonathan Hui. [RL — Meta-Learning](https://jonathan-hui.medium.com/meta-learning-how-we-address-the-shortcomings-of-our-deep-networks-a008aa4b5b2b).

**Comment by NorbertZheng** [comment]
## Overview
So far, AI cannot learn as efficiently as a human. Many classifiers take millions of training samples to train and knowledge is not shared among tasks. Each task is trained independently from others. In this article, we will look into the problems and then check out some proposed solutions.

**What is the problem?**
Compared with a human, there are two critical weaknesses:
 - Sample efficiency: Deep learning has poor sample efficiency. For example, to recognize digit handwriting, we typically read 6000 samples per digit.
 - Poor transferability. We don’t learn from previous experience or knowledge.

**Meta-Learning**
So what is meta-learning? We define it as “learning how to learn”. Personally, I don’t think this clarifies anything. In reality, we don’t know the exact definition or the solution yet. So
 - meta-learning is still a loose term referring to different approaches.

In this article, we will focus on the following areas:
 - recurrent models,
 - meta-optimization,
 - and metric learning.

**Comment by NorbertZheng** [comment]
## Few-Shot Learning
But let’s define some basic concepts first. In CIFAR-10, we have 60,000 images in ten different categories. In few-shot learning, we
 - train a model with a far larger amount of tasks but **far fewer samples per task**.

Our ultimate goal is to generalize knowledge and apply the model to new tasks that **even not trained before**.

For example, in task 1, we are given 3 samples to learn emoji. Then, with a new emoji, we train the model to associate it with one of the previous samples. E.g. In-Context Learning?
![image](https://user-images.githubusercontent.com/37235023/218360079-7d9d12e8-bfce-4558-a317-f73f9751c5ce.png)

In our second task, we train it with the alphabets.
![image](https://user-images.githubusercontent.com/37235023/218360131-354d8d19-d96a-4535-b62a-5ea3512d67c9.png)

We repeat the process many times with different tasks. Once the training is done, we measure the model's versatility by testing it with a task that we have not performed before, distinguishing the Chinese characters. And yet, the model can associate the testing samples correctly with the new input.
![image](https://user-images.githubusercontent.com/37235023/218360412-433ab36d-bcd6-4966-b1c5-8fa07c072bb4.png)

**Comment by NorbertZheng** [comment]
We may wonder what is the difference between few-shot training and our traditional DL using a large dataset. In DL, we ~~use regularization to make sure we are not overfitting our model with a small dataset~~. But by training the model with so many samples and iterations, we are overfitting our tasks.
 - What we learned cannot be generalized to other tasks.

Let me demonstrate some of the constraints in DL. We often get stuck when we test samples that are not common in the dataset. For example, in the toy category, there are way too many variants that a DL model can learn. As shown below, the yellow toy duck is badly classified.
 - In few-shot training, the key objective is to process data that we have not been trained in before.

![image](https://user-images.githubusercontent.com/37235023/218360801-a7cfe2c0-04d4-4b87-aa79-42228eb253c2.png)
Mistaken the duck toy as a savings bank or a musical instrument.

**Comment by NorbertZheng** [comment]
### One-Shot Learning
In a one-shot training, we will only provide one training sample per category. In the example below, the training contains multiple datasets. Each dataset contains a 1-shot, 5-class classification task, i.e. five samples from five different classes. Then the next dataset is for 5 different categories.

![image](https://user-images.githubusercontent.com/37235023/218361298-a84bd829-a50d-40dd-8321-cb8bbc92cc6b.png)
Modified from [source](https://openreview.net/pdf?id=rJY0-Kcll).

In this one-shot training, we can train an RNN to learn the training data and labels. When we present a test input, we should predict its label correctly. (We will come back on this later.) We should note that this is still in the Meta-Train process, we use the error between the predicted label and the target label to train the model.

![image](https://user-images.githubusercontent.com/37235023/218364215-944c2ccd-87ba-4f1c-a7b8-1d0b308e625b.png)
[Source](http://rail.eecs.berkeley.edu/deeprlcourse-fa17/f17docs/lecture_16_meta_learning.pdf).

In meta-testing, we provide datasets that have never been trained before. In this example, the key focus for meta-learning is to learn the secret of classifying objects. Once we have learned from hundred tasks, we would not focus on individual classes. Instead, we should
 - discover the general knowledge in classifying objects.

So even when we are presented with classes never seen before, we should manage to solve the problem.

**Comment by NorbertZheng** [comment]
If we collect tasks smarter, we learn better. Still task-oriented.

**Comment by NorbertZheng** [comment]
### Omniglot
Before we go into the details. Let’s introduce Omniglot. It is a popular dataset for the few-shot learning. The following are 20 drawings from Omniglot that represent 20 classes.

![image](https://user-images.githubusercontent.com/37235023/218365173-abb65168-8feb-43a2-9fe7-b59e90c3d031.png)

**Comment by NorbertZheng** [comment]
## Recurrent Models
The first meta-learning approach is the recurrent model. We feed data into an RNN-like model to remember what we see so far. When we are present with a testing input, we **recall from memory to tell us what it is**, e.g. In-Context Learning (ICL). However, we don’t have enough memory for everything we see. The recurrent model stores features and we use linear algebra similar to word embedding to correlate information.

![image](https://user-images.githubusercontent.com/37235023/218365505-a4580d85-9699-40b8-9122-29dc2104a6d6.png)

### Memory Network (MN)
Let’s do a recap in a memory network (MN) first. MN uses a controller to extract features from the input. Then we use the features to access memory.

![image](https://user-images.githubusercontent.com/37235023/218387232-93ac3b70-550c-40e1-a404-0c8890955b9d.png)

For example, you take a phone call but you cannot recognize the voice immediately. The voice sounds a whole lot like your cousin (0.7 chance) but it also resembles the voice of your elder brother (0.3 chance). In the diagram above, each row represents an object. We compute a weight $w$ for each row to measure its relevancy with the input. Then we computed a weighted sum from all rows as the memory output. For classification problems, we can fit the output to a classifier to predict its class.

![image](https://user-images.githubusercontent.com/37235023/218387843-d6250170-e7f3-42c2-8f31-1612b91c02b0.png)
(If you want more details, take a quick look at [Neural Turing Machine](https://medium.com/@jonathan_hui/neural-turing-machines-a-fundamental-approach-to-access-memory-in-deep-learning-b823a31fe91d) later.)

**Comment by NorbertZheng** [comment]
### Memory-Augmented Neural Networks
Memory-Augmented Neural Networks are
 - one of the meta-learning methods using an external memory network with RNN.

In supervised learning, we provide both input and label at the same time in step $t$. However, in this model, the label is not provided until the next timestamp $t+1$ (shown below). This is a technique to
 - discourage the RNN cell to map the input to the class label directly.

We want our model to memorize the experience.

![image](https://user-images.githubusercontent.com/37235023/218388638-17cb962c-b3e5-4e45-8481-693a168cdc83.png)

Training Memory-Augmented Neural Networks.
![image](https://user-images.githubusercontent.com/37235023/218389138-8658c2fb-0f85-43b9-94db-26962e72c1bd.png)
Modified from [source](https://arxiv.org/pdf/1605.06065.pdf).

In Memory-Augmented Neural Networks, we use external memory to store sample representation and class label information.
 - A controller, typically implemented as LSTM, produces a key from the input which is either stored in the external memory or used to retrieve a particular memory.

The whole system is then trained with backpropagation.

![image](https://user-images.githubusercontent.com/37235023/218389539-b5067f32-e68b-4017-9278-0de5c48ed002.png)

After the first episode, the samples will be shuffled and the model will start establishing memory in predicting labels, which is pretty much like the process in TEM switching between environments.

![image](https://user-images.githubusercontent.com/37235023/218389834-b0555907-1b5f-4281-b78b-2ddc9c264c6a.png)
[source](https://arxiv.org/pdf/1605.06065.pdf).

For details, we recommend readers to read the original [paper](https://arxiv.org/pdf/1605.06065.pdf).

 - If we can learn from experience, we learn better.

**Comment by NorbertZheng** [comment]
## Learning optimizers
In the second meta-learning approach, we optimize the model more efficiently. After the training of each task, we can use the information to update the model.

$$
\theta \leftarrow \theta + \alpha \nabla_{\theta}R(\theta).
$$

Yet, we are learning a particular task rather than finding the fundamental knowledge behind all the learned tasks. So
 - instead of updating the model immediately, we wait until a batch of tasks is completed.

We later merge all we learned from these tasks for a single update. We combine the reward gradient for each newly suggested parameters learned from each task. Then we perform the backpropagation. This approach fulfills the concept of “learn what we learn”.

![image](https://user-images.githubusercontent.com/37235023/218391232-91509c1d-f47d-483f-a02f-e5c44a86e931.png)

**Comment by NorbertZheng** [comment]
### MAML
The Model-agnostic meta-learning (MAML) utilizes the concept above to update models. It is simple and it is almost the same as our traditional DL gradient descent with one added line of code. Here, we do not update the model parameters immediately after each task. Instead, we wait until a batch of tasks to be completed.

![image](https://user-images.githubusercontent.com/37235023/218392135-859e26b0-df28-45a4-b84f-9ad38cc3ffc6.png)
[Source](https://arxiv.org/pdf/1703.03400.pdf).

where $\theta_{i}'$ is the suggested model updates from task $T_{i}$, and $\theta$ is our final model updates.

For each task, we use backpropagation to compute a suggested model.

$$
\theta_{i}'=\theta-\alpha\nabla_{\theta}L_{T_{i}}(f_{\theta}).
$$

We then consolidate the losses of the trained tasks and backpropagate the losses to make the next model update:

$$
\theta \leftarrow \theta - \beta\nabla_{\theta}\Sigma_{T_{i}\sim  p(T)}L_{T_{i}}(f_{\theta_{i}'}).
$$

But in this step, we sample new task $T_{i}$ from $p(T)$ to compute the loss. Conceptually, we are finding a model that minimizes the losses of the tasks.

![image](https://user-images.githubusercontent.com/37235023/218393084-40f64854-0fa2-4486-b8b0-e38c7eea677f.png)
[Source](https://arxiv.org/pdf/1703.03400.pdf).

Graphically, each task may drive the model parameters in different directions. Introducing the meta-learning step and the few-shot datasets, we learn a model that works with many tasks but not overfit with a particular one.

**Comment by NorbertZheng** [comment]
There are other optimizers with the objective of learning more efficiently. For example, OpenAI proposes another optimizer call Reptile. In stochastic gradient descent, we compute one gradient descent and update the model. Then we fetch the next batch of data for the next iteration. In Reptile, it performs multiple-step gradient descent for each task and uses the result in the last step to update the model with a concept similar to the running mean.

![image](https://user-images.githubusercontent.com/37235023/218393326-878f99da-b82b-4746-aa0f-d4aef704d52a.png)

In the OpenAI paper, it demonstrates mathematically why both MAML and Reptile behave similarly. In short, it argues that MAML uses.

But we will let you to read the original paper for the arguments.

 - If we optimize better, we learn better.

**Comment by NorbertZheng** [comment]
## Metric learning (Representation Learning?)
The third meta-learning approach we will discuss is the metric learning. Do you remember picture pixel-by-pixel? No. To learn, we need to extract the maximum information we care with the least amount of memory. So, the third meta-learning approach focuses on
 - how well we extract features but not overdo it.

### Siamese Neural Networks
In Siamese Neural Networks (the figure below), we use two identical networks sharing the same model parameter values to extract features for two samples. Then we feed the extracted features into a discriminator to identify whether both samples belong to the same class of object. For example, we can compute its cosine similarity (p) of their feature vectors. If they are similar, p should be close to 1. Otherwise, they should be close to 0. Based on the labels of the samples and p, we train the networks accordingly. In short, we want to find features that make samples belong to the same class or differentiate them.

![image](https://user-images.githubusercontent.com/37235023/218394058-af010686-b12c-4114-92e1-6891cb117e74.png)
Contrastive Learning. A sub-field of Representation Learning.

**Comment by NorbertZheng** [comment]
### Matching Network
There is another method called Matching Network that is very similar to the Siamese Neural Networks.

![image](https://user-images.githubusercontent.com/37235023/218395159-efe5e208-cdf1-41c3-a138-e26e7a16ddda.png)
[Source](https://arxiv.org/pdf/1606.04080.pdf).

$g$ and $f$ are feature extractors using deep networks to extract features for our input and our testing (the German Shepard) samples. Usually, $g$ and $f$ are the same and share the same deep network. Then we compare their similarity and use a softmax function to compute whether they are similar. Once again, we compute a cost function from our prediction to train our feature extractor. Here is the mathematical formalization:

$$
\begin{aligned}
\hat{y}&=\sum_{i=1}^{k}a(\hat{x},x_{i})y_{i},\\
a(\hat{x},x_{i})&=\frac{e^{c(f(\hat{x}),g(x_{i}))}}{\Sigma_{jj=1}^{k}e^{c(f(\hat{x}),g(x_{j}))}},\\
\end{aligned}
$$

where $\hat{x}$ is our input testing image, $x_{i}$ is our samples $$i$, $y_{i}$ is its label, $\hat{y}$ is our class prediction.

 - If we know how to represent data better, we learn better.

**Comment by NorbertZheng** [comment]
## Other approaches
There are other meta-learning approaches focus on how to tune a model with better [hyperparameters](https://arxiv.org/pdf/1502.03492.pdf). i.e. Learn how to tune the hyperparameters smartly. Alternatively, we can combine DL layers to form a new model dynamically.

![image](https://user-images.githubusercontent.com/37235023/218398481-e79cff15-57ce-4222-8aff-8bdc2d71341d.png)
[source](https://arxiv.org/pdf/1611.01578.pdf).

These methods make models more accurate but not necessary more efficient in learning with fewer samples. So we will not discuss further in our meta-learning discussion.

**Comment by NorbertZheng** [comment]
## Thoughts
Learn how to learn better is not a challenge for the machines only but for the human also. Meta-learning has been studied for decades but yet we do not fully understand how it is done. To close our thoughts, here is the possible area of study to improve learning efficiency.
 - Collect better information to learn.
 - Learn from past experiences better.
 - Know better how to represent information.
 - How to optimize (solve) model better.
 - Explore better.
 - Associate better.
 - Generalize better.
